File: fraiseql-go/fraiseql/custom_scalar_test.go
package fraiseql

import (
	"fmt"
	"strings"
	"testing"
)

// EmailScalar is a test email scalar implementation
type EmailScalar struct{}

func (e *EmailScalar) Name() string {
	return "Email"
}

func (e *EmailScalar) Serialize(value interface{}) (interface{}, error) {
	return toString(value), nil
}

func (e *EmailScalar) ParseValue(value interface{}) (interface{}, error) {
	str := toString(value)
	str = strings.TrimSpace(str)
	if !strings.Contains(str, "@") {
		return nil, fmt.Errorf("invalid email format: %s", str)
	}
	return str, nil
}

func (e *EmailScalar) ParseLiteral(ast interface{}) (interface{}, error) {
	m, ok := ast.(map[string]interface{})
	if ok {
		if val, exists := m["value"]; exists {
			return e.ParseValue(val)
		}
	}
	return nil, fmt.Errorf("email literal must be string")
}

// PhoneScalar is a test phone scalar implementation
type PhoneScalar struct{}

func (p *PhoneScalar) Name() string {
	return "Phone"
}

func (p *PhoneScalar) Serialize(value interface{}) (interface{}, error) {
	return toString(value), nil
}

func (p *PhoneScalar) ParseValue(value interface{}) (interface{}, error) {
	str := toString(value)
	str = strings.TrimSpace(str)
	if !strings.HasPrefix(str, "+") {
		return nil, fmt.Errorf("phone must start with +")
	}
	digits := str[1:]
	for _, c := range digits {
		if c < '0' || c > '9' {
			return nil, fmt.Errorf("phone must contain only digits after +")
		}
	}
	if len(digits) < 10 || len(digits) > 14 {
		return nil, fmt.Errorf("phone must be 10-14 digits, got %d", len(digits))
	}
	return str, nil
}

func (p *PhoneScalar) ParseLiteral(ast interface{}) (interface{}, error) {
	m, ok := ast.(map[string]interface{})
	if ok {
		if val, exists := m["value"]; exists {
			return p.ParseValue(val)
		}
	}
	return nil, fmt.Errorf("phone literal must be string")
}

// ========================================================================
// Scalar Registration Tests
// ========================================================================

func TestRegisterCustomScalarGlobally(t *testing.T) {
	defer Reset()

	// Before registration
	if HasCustomScalar("Email") {
		t.Fatal("Email should not be registered yet")
	}

	// Register Email
	RegisterCustomScalar(&EmailScalar{})

	// After registration
	if !HasCustomScalar("Email") {
		t.Fatal("Email should be registered")
	}

	scalar := GetCustomScalar("Email")
	if scalar == nil {
		t.Fatal("Email scalar should be retrievable")
	}
}

func TestPreventsDuplicateNames(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})

	defer func() {
		if r := recover(); r == nil {
			t.Fatal("Expected panic for duplicate scalar name")
		}
	}()

	// Try to register another Email scalar
	RegisterCustomScalar(&EmailScalar{})
}

func TestRegistersMultipleScalars(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})
	RegisterCustomScalar(&PhoneScalar{})

	if !HasCustomScalar("Email") || !HasCustomScalar("Phone") {
		t.Fatal("Both scalars should be registered")
	}
}

func TestValidatesScalarHasName(t *testing.T) {
	defer Reset()

	// Create a scalar with empty name
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("Expected panic for empty scalar name")
		}
	}()

	RegisterCustomScalar(&NoNameScalar{})
}

// NoNameScalar is a test scalar with empty name (for testing validation)
type NoNameScalar struct{}

func (n *NoNameScalar) Name() string {
	return ""
}

func (n *NoNameScalar) Serialize(v interface{}) (interface{}, error) {
	return v, nil
}

func (n *NoNameScalar) ParseValue(v interface{}) (interface{}, error) {
	return v, nil
}

func (n *NoNameScalar) ParseLiteral(v interface{}) (interface{}, error) {
	return v, nil
}

// ========================================================================
// Validation Engine Tests
// ========================================================================

func TestValidateParseValueSuccess(t *testing.T) {
	email := &EmailScalar{}
	result, err := ValidateCustomScalar(email, "user@example.com", "parseValue")
	if err != nil {
		t.Fatalf("Validation should succeed: %v", err)
	}
	if result != "user@example.com" {
		t.Fatalf("Expected 'user@example.com', got %v", result)
	}
}

func TestValidateParseValueFailure(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "invalid-email", "parseValue")
	if err == nil {
		t.Fatal("Validation should fail for invalid email")
	}

	valErr, ok := err.(*ScalarValidationError)
	if !ok {
		t.Fatal("Error should be ScalarValidationError")
	}
	if valErr.ScalarName != "Email" {
		t.Fatalf("Expected scalar name 'Email', got %s", valErr.ScalarName)
	}
}

func TestValidateSerialize(t *testing.T) {
	email := &EmailScalar{}
	result, err := ValidateCustomScalar(email, "user@example.com", "serialize")
	if err != nil {
		t.Fatalf("Serialization should succeed: %v", err)
	}
	if result != "user@example.com" {
		t.Fatalf("Expected 'user@example.com', got %v", result)
	}
}

func TestValidateParseLiteral(t *testing.T) {
	email := &EmailScalar{}
	ast := map[string]interface{}{"value": "user@example.com"}
	result, err := ValidateCustomScalar(email, ast, "parseLiteral")
	if err != nil {
		t.Fatalf("Parsing literal should succeed: %v", err)
	}
	if result != "user@example.com" {
		t.Fatalf("Expected 'user@example.com', got %v", result)
	}
}

func TestValidateMultipleScalars(t *testing.T) {
	defer Reset()

	email := &EmailScalar{}
	phone := &PhoneScalar{}

	// Email validation
	result, err := ValidateCustomScalar(email, "test@test.com", "parseValue")
	if err != nil {
		t.Fatalf("Email validation should succeed: %v", err)
	}
	if result != "test@test.com" {
		t.Fatalf("Expected 'test@test.com', got %v", result)
	}

	// Phone validation
	result, err = ValidateCustomScalar(phone, "+12025551234", "parseValue")
	if err != nil {
		t.Fatalf("Phone validation should succeed: %v", err)
	}
	if result != "+12025551234" {
		t.Fatalf("Expected '+12025551234', got %v", result)
	}

	// Both fail appropriately
	_, err = ValidateCustomScalar(email, "notanemail", "parseValue")
	if err == nil {
		t.Fatal("Email validation should fail")
	}

	_, err = ValidateCustomScalar(phone, "invalid", "parseValue")
	if err == nil {
		t.Fatal("Phone validation should fail")
	}
}

func TestValidateInvalidContext(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "test@test.com", "invalidContext")
	if err == nil {
		t.Fatal("Should fail for invalid context")
	}
}

func TestValidateDefaultContext(t *testing.T) {
	email := &EmailScalar{}
	result, err := ValidateCustomScalarWithDefault(email, "test@test.com")
	if err != nil {
		t.Fatalf("Validation with default context should succeed: %v", err)
	}
	if result != "test@test.com" {
		t.Fatalf("Expected 'test@test.com', got %v", result)
	}
}

// ========================================================================
// Error Message Tests
// ========================================================================

func TestErrorIncludesScalarName(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "notanemail", "parseValue")
	if err == nil {
		t.Fatal("Should have error")
	}

	if !strings.Contains(err.Error(), "Email") {
		t.Fatalf("Error should include scalar name: %s", err.Error())
	}
}

func TestErrorIncludesContext(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "notanemail", "parseValue")
	if err == nil {
		t.Fatal("Should have error")
	}

	if !strings.Contains(err.Error(), "parseValue") {
		t.Fatalf("Error should include context: %s", err.Error())
	}
}

func TestErrorIncludesMessage(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "notanemail", "parseValue")
	if err == nil {
		t.Fatal("Should have error")
	}

	if !strings.Contains(err.Error(), "invalid email format") {
		t.Fatalf("Error should include message: %s", err.Error())
	}
}

func TestErrorGetters(t *testing.T) {
	email := &EmailScalar{}
	_, err := ValidateCustomScalar(email, "notanemail", "parseValue")
	if err == nil {
		t.Fatal("Should have error")
	}

	valErr, ok := err.(*ScalarValidationError)
	if !ok {
		t.Fatal("Error should be ScalarValidationError")
	}

	if valErr.ScalarName != "Email" {
		t.Fatalf("Expected scalar name 'Email', got %s", valErr.ScalarName)
	}

	if valErr.Context != "parseValue" {
		t.Fatalf("Expected context 'parseValue', got %s", valErr.Context)
	}
}

// ========================================================================
// Utility Function Tests
// ========================================================================

func TestGetAllCustomScalars(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})
	RegisterCustomScalar(&PhoneScalar{})

	scalars := GetAllCustomScalars()

	if len(scalars) != 2 {
		t.Fatalf("Expected 2 scalars, got %d", len(scalars))
	}

	if _, ok := scalars["Email"]; !ok {
		t.Fatal("Email should be in scalars")
	}

	if _, ok := scalars["Phone"]; !ok {
		t.Fatal("Phone should be in scalars")
	}
}

func TestGetAllCustomScalarsEmpty(t *testing.T) {
	defer Reset()

	scalars := GetAllCustomScalars()
	if len(scalars) != 0 {
		t.Fatalf("Expected 0 scalars, got %d", len(scalars))
	}
}

// ========================================================================
// Email Scalar Specific Tests
// ========================================================================

func TestEmailValidatesCorrect(t *testing.T) {
	email := &EmailScalar{}
	result, err := email.ParseValue("test@example.com")
	if err != nil {
		t.Fatalf("Should validate correct email: %v", err)
	}
	if result != "test@example.com" {
		t.Fatalf("Expected 'test@example.com', got %v", result)
	}
}

func TestEmailRejectsNoAt(t *testing.T) {
	email := &EmailScalar{}
	_, err := email.ParseValue("notanemail")
	if err == nil {
		t.Fatal("Should reject email without @")
	}
}

func TestEmailSerializes(t *testing.T) {
	email := &EmailScalar{}
	result, err := email.Serialize("test@example.com")
	if err != nil {
		t.Fatalf("Should serialize: %v", err)
	}
	if result != "test@example.com" {
		t.Fatalf("Expected 'test@example.com', got %v", result)
	}
}

// ========================================================================
// Phone Scalar Specific Tests
// ========================================================================

func TestPhoneValidatesE164(t *testing.T) {
	phone := &PhoneScalar{}
	result, err := phone.ParseValue("+12025551234")
	if err != nil {
		t.Fatalf("Should validate E.164 format: %v", err)
	}
	if result != "+12025551234" {
		t.Fatalf("Expected '+12025551234', got %v", result)
	}
}

func TestPhoneRequiresPlus(t *testing.T) {
	phone := &PhoneScalar{}
	_, err := phone.ParseValue("2025551234")
	if err == nil {
		t.Fatal("Should require + prefix")
	}
}

func TestPhoneValidatesDigitCount(t *testing.T) {
	phone := &PhoneScalar{}
	_, err := phone.ParseValue("+123")
	if err == nil {
		t.Fatal("Should validate digit count")
	}
}

func TestPhoneSerializes(t *testing.T) {
	phone := &PhoneScalar{}
	result, err := phone.Serialize("+12025551234")
	if err != nil {
		t.Fatalf("Should serialize: %v", err)
	}
	if result != "+12025551234" {
		t.Fatalf("Expected '+12025551234', got %v", result)
	}
}

// ========================================================================
// Schema Export Tests
// ========================================================================

func TestCustomScalarsExportedInSchema(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})
	RegisterCustomScalar(&PhoneScalar{})

	schema := GetSchema()

	if len(schema.CustomScalars) != 2 {
		t.Fatalf("Expected 2 custom scalars in schema, got %d", len(schema.CustomScalars))
	}

	// Check that scalars are in the schema
	hasEmail := false
	hasPhone := false

	for _, scalar := range schema.CustomScalars {
		name, ok := scalar["name"].(string)
		if ok {
			if name == "Email" {
				hasEmail = true
			}
			if name == "Phone" {
				hasPhone = true
			}
		}
	}

	if !hasEmail {
		t.Fatal("Email scalar should be in schema")
	}
	if !hasPhone {
		t.Fatal("Phone scalar should be in schema")
	}
}

func TestUnregisterCustomScalar(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})
	if !HasCustomScalar("Email") {
		t.Fatal("Email should be registered")
	}

	UnregisterCustomScalar("Email")
	if HasCustomScalar("Email") {
		t.Fatal("Email should be unregistered")
	}
}

func TestClearCustomScalars(t *testing.T) {
	defer Reset()

	RegisterCustomScalar(&EmailScalar{})
	RegisterCustomScalar(&PhoneScalar{})

	ClearCustomScalars()

	if len(GetAllCustomScalars()) != 0 {
		t.Fatal("All scalars should be cleared")
	}
}

File: sdks/official/fraiseql-java/src/test/java/com/fraiseql/core/CustomScalarTest.java
package com.fraiseql.core;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;

/**
 * Comprehensive test suite for custom scalar support.
 *
 * <p>Tests cover registration, validation, error handling, and integration with schema.
 */
@DisplayName("Custom Scalar Support")
class CustomScalarTest {

    /**
     * Test Email scalar implementation.
     */
    @Scalar
    static class Email extends CustomScalar {
        @Override
        public String getName() {
            return "Email";
        }

        @Override
        public Object serialize(Object value) {
            return String.valueOf(value);
        }

        @Override
        public Object parseValue(Object value) {
            String str = String.valueOf(value).trim();
            if (!str.contains("@")) {
                throw new IllegalArgumentException("Invalid email format: " + str);
            }
            return str;
        }

        @Override
        public Object parseLiteral(Object ast) {
            if (ast instanceof java.util.Map) {
                java.util.Map<String, Object> astMap = (java.util.Map<String, Object>) ast;
                if (astMap.containsKey("value")) {
                    return parseValue(astMap.get("value"));
                }
            }
            throw new IllegalArgumentException("Email literal must be string");
        }
    }

    /**
     * Test Phone scalar implementation.
     */
    @Scalar
    static class Phone extends CustomScalar {
        @Override
        public String getName() {
            return "Phone";
        }

        @Override
        public Object serialize(Object value) {
            return String.valueOf(value);
        }

        @Override
        public Object parseValue(Object value) {
            String str = String.valueOf(value).trim();
            if (!str.startsWith("+")) {
                throw new IllegalArgumentException("Phone must start with +");
            }
            String digits = str.substring(1);
            if (!digits.matches("^\\d+$")) {
                throw new IllegalArgumentException("Phone must contain only digits after +");
            }
            if (digits.length() < 10 || digits.length() > 14) {
                throw new IllegalArgumentException(
                    String.format("Phone must be 10-14 digits, got %d", digits.length()));
            }
            return str;
        }

        @Override
        public Object parseLiteral(Object ast) {
            if (ast instanceof java.util.Map) {
                java.util.Map<String, Object> astMap = (java.util.Map<String, Object>) ast;
                if (astMap.containsKey("value")) {
                    return parseValue(astMap.get("value"));
                }
            }
            throw new IllegalArgumentException("Phone literal must be string");
        }
    }

    @BeforeEach
    void setUp() {
        ScalarProcessor.clearAll();
    }

    // ========================================================================
    // Scalar Registration Tests
    // ========================================================================

    @Test
    @DisplayName("Registers scalar globally")
    void testScalarRegistersGlobally() {
        // Before registration
        assertTrue(ScalarRegistry.getInstance().getCustomScalars().isEmpty());

        // Register Email
        ScalarProcessor.register(Email.class);

        // After registration
        assertTrue(ScalarRegistry.getInstance().hasScalar("Email"));
        assertEquals(Email.class, ScalarRegistry.getInstance().getScalar("Email"));
    }

    @Test
    @DisplayName("Validates class has @Scalar annotation")
    void testValidatesScalarAnnotation() {
        class NotAnnotated extends CustomScalar {
            @Override
            public String getName() {
                return "NotAnnotated";
            }

            @Override
            public Object serialize(Object value) {
                return value;
            }

            @Override
            public Object parseValue(Object value) {
                return value;
            }

            @Override
            public Object parseLiteral(Object ast) {
                return ast;
            }
        }

        assertThrows(IllegalArgumentException.class, () -> ScalarProcessor.register(NotAnnotated.class));
    }

    @Test
    @DisplayName("Validates scalar name is non-empty")
    void testValidatesScalarName() {
        @Scalar
        class NoName extends CustomScalar {
            @Override
            public String getName() {
                return "";
            }

            @Override
            public Object serialize(Object value) {
                return value;
            }

            @Override
            public Object parseValue(Object value) {
                return value;
            }

            @Override
            public Object parseLiteral(Object ast) {
                return ast;
            }
        }

        assertThrows(IllegalArgumentException.class, () -> ScalarProcessor.register(NoName.class));
    }

    @Test
    @DisplayName("Prevents duplicate scalar names")
    void testPreventsDuplicateNames() {
        ScalarProcessor.register(Email.class);

        @Scalar
        class AnotherEmail extends CustomScalar {
            @Override
            public String getName() {
                return "Email";
            }

            @Override
            public Object serialize(Object value) {
                return value;
            }

            @Override
            public Object parseValue(Object value) {
                return value;
            }

            @Override
            public Object parseLiteral(Object ast) {
                return ast;
            }
        }

        assertThrows(IllegalArgumentException.class, () -> ScalarProcessor.register(AnotherEmail.class));
    }

    @Test
    @DisplayName("Registers multiple scalars")
    void testRegistersMultipleScalars() {
        ScalarProcessor.registerAll(Email.class, Phone.class);

        assertTrue(ScalarRegistry.getInstance().hasScalar("Email"));
        assertTrue(ScalarRegistry.getInstance().hasScalar("Phone"));
    }

    // ========================================================================
    // Validation Engine Tests
    // ========================================================================

    @Test
    @DisplayName("Validates parseValue successfully")
    void testValidateParseValueSuccess() {
        ScalarProcessor.register(Email.class);
        Object result = ScalarValidator.validate(Email.class, "user@example.com", "parseValue");
        assertEquals("user@example.com", result);
    }

    @Test
    @DisplayName("Rejects invalid values")
    void testValidateParseValueFailure() {
        ScalarProcessor.register(Email.class);
        assertThrows(
            ScalarValidationError.class,
            () -> ScalarValidator.validate(Email.class, "invalid-email", "parseValue"));
    }

    @Test
    @DisplayName("Serializes values")
    void testValidateSerialize() {
        ScalarProcessor.register(Email.class);
        Object result = ScalarValidator.validate(Email.class, "user@example.com", "serialize");
        assertEquals("user@example.com", result);
    }

    @Test
    @DisplayName("Parses literals")
    void testValidateParseLiteral() {
        ScalarProcessor.register(Email.class);
        java.util.Map<String, Object> ast = java.util.Collections.singletonMap("value", "user@example.com");
        Object result = ScalarValidator.validate(Email.class, ast, "parseLiteral");
        assertEquals("user@example.com", result);
    }

    @Test
    @DisplayName("Handles multiple scalar types")
    void testValidateMultipleScalars() {
        ScalarProcessor.registerAll(Email.class, Phone.class);

        // Email validation
        Object email = ScalarValidator.validate(Email.class, "test@test.com");
        assertEquals("test@test.com", email);

        // Phone validation
        Object phone = ScalarValidator.validate(Phone.class, "+12025551234");
        assertEquals("+12025551234", phone);

        // Both fail appropriately
        assertThrows(ScalarValidationError.class, () -> ScalarValidator.validate(Email.class, "notanemail"));
        assertThrows(ScalarValidationError.class, () -> ScalarValidator.validate(Phone.class, "invalid"));
    }

    @Test
    @DisplayName("Throws on invalid context")
    void testValidateInvalidContext() {
        ScalarProcessor.register(Email.class);
        assertThrows(
            IllegalArgumentException.class,
            () -> ScalarValidator.validate(Email.class, "test@test.com", "invalidContext"));
    }

    @Test
    @DisplayName("Uses default context parseValue")
    void testValidateDefaultContext() {
        ScalarProcessor.register(Email.class);
        Object result = ScalarValidator.validate(Email.class, "test@test.com");
        assertEquals("test@test.com", result);
    }

    // ========================================================================
    // Error Message Tests
    // ========================================================================

    @Test
    @DisplayName("Error includes scalar name")
    void testErrorIncludesScalarName() {
        ScalarProcessor.register(Email.class);
        ScalarValidationError error =
            assertThrows(
                ScalarValidationError.class,
                () -> ScalarValidator.validate(Email.class, "notanemail"));
        assertTrue(error.getMessage().contains("Email"));
    }

    @Test
    @DisplayName("Error includes context")
    void testErrorIncludesContext() {
        ScalarProcessor.register(Email.class);
        ScalarValidationError error =
            assertThrows(
                ScalarValidationError.class,
                () -> ScalarValidator.validate(Email.class, "notanemail", "parseValue"));
        assertTrue(error.getMessage().contains("parseValue"));
    }

    @Test
    @DisplayName("Error includes original message")
    void testErrorIncludesMessage() {
        ScalarProcessor.register(Email.class);
        ScalarValidationError error =
            assertThrows(
                ScalarValidationError.class,
                () -> ScalarValidator.validate(Email.class, "notanemail"));
        assertTrue(error.getMessage().contains("Invalid email format"));
    }

    @Test
    @DisplayName("Error getters work")
    void testErrorGetters() {
        ScalarProcessor.register(Email.class);
        ScalarValidationError error =
            assertThrows(
                ScalarValidationError.class,
                () -> ScalarValidator.validate(Email.class, "notanemail", "parseValue"));

        assertEquals("Email", error.getScalarName());
        assertEquals("parseValue", error.getContext());
    }

    // ========================================================================
    // Utility Function Tests
    // ========================================================================

    @Test
    @DisplayName("Gets all custom scalars")
    void testGetAllCustomScalars() {
        ScalarProcessor.registerAll(Email.class, Phone.class);

        java.util.Map<String, Class<? extends CustomScalar>> scalars =
            ScalarValidator.getAllCustomScalars();
        assertTrue(scalars.containsKey("Email"));
        assertTrue(scalars.containsKey("Phone"));
        assertEquals(Email.class, scalars.get("Email"));
        assertEquals(Phone.class, scalars.get("Phone"));
    }

    @Test
    @DisplayName("Returns empty when no scalars registered")
    void testGetAllCustomScalarsEmpty() {
        java.util.Map<String, Class<? extends CustomScalar>> scalars =
            ScalarValidator.getAllCustomScalars();
        assertTrue(scalars.isEmpty());
    }

    // ========================================================================
    // Email Scalar Specific Tests
    // ========================================================================

    @Test
    @DisplayName("Email validates correct addresses")
    void testEmailValidatesCorrect() {
        Email email = new Email();
        Object result = email.parseValue("test@example.com");
        assertEquals("test@example.com", result);
    }

    @Test
    @DisplayName("Email rejects without @")
    void testEmailRejectsNoAt() {
        Email email = new Email();
        assertThrows(IllegalArgumentException.class, () -> email.parseValue("notanemail"));
    }

    @Test
    @DisplayName("Email serializes")
    void testEmailSerializes() {
        Email email = new Email();
        Object result = email.serialize("test@example.com");
        assertEquals("test@example.com", result);
    }

    // ========================================================================
    // Phone Scalar Specific Tests
    // ========================================================================

    @Test
    @DisplayName("Phone validates E.164 format")
    void testPhoneValidatesE164() {
        Phone phone = new Phone();
        Object result = phone.parseValue("+12025551234");
        assertEquals("+12025551234", result);
    }

    @Test
    @DisplayName("Phone requires + prefix")
    void testPhoneRequiresPlus() {
        Phone phone = new Phone();
        assertThrows(IllegalArgumentException.class, () -> phone.parseValue("2025551234"));
    }

    @Test
    @DisplayName("Phone validates digit count")
    void testPhoneValidatesDigitCount() {
        Phone phone = new Phone();
        assertThrows(IllegalArgumentException.class, () -> phone.parseValue("+123"));
    }

    @Test
    @DisplayName("Phone serializes")
    void testPhoneSerializes() {
        Phone phone = new Phone();
        Object result = phone.serialize("+12025551234");
        assertEquals("+12025551234", result);
    }

    // ========================================================================
    // Annotation Tests
    // ========================================================================

    @Test
    @DisplayName("@Scalar annotation is present on Email")
    void testScalarAnnotationPresent() {
        assertTrue(Email.class.isAnnotationPresent(Scalar.class));
    }

    @Test
    @DisplayName("CustomScalar toString works")
    void testToString() {
        Email email = new Email();
        assertEquals("CustomScalar(Email)", email.toString());
    }
}
